code/go/0chain.net/smartcontract/storagesc/blobber.go

```diff
@@ -31,6 +31,10 @@ func getBlobber(
 	if err != nil {
 		return nil, err
 	}
+	if blobber.ProviderType != spenum.Blobber {
```

Comment:
It's impossible to get a storageNode from MPT that has different provider type other than `spenum.Blobber`, if we want to check if the same ID is used for other providers, we will then need to check `ValidatorNode`, `MinerNode`, `AuthorizerNode`, etc. But that will make the code ugly. Anyway we can do it in the following way
1. Define a `Provider` struct in the `provider` package with struct like:
```go
type Provider struct {
    ID string `json:"id"`
    ProviderType spenum.Provider `json:"provider_type"`
}
```
2. Embed the `Provider` to all of those `StorageNode`, `ValidatorNode`...
3. When need to check provider type, we can do:
```go
var p provider.Provider
balances.GetTrieNode(provider.GetKey(id), &p)
```
Then check if the provider type is the correct one.

Reply:
My aim was to reduce the number of MPT lookups. 

If I understand correctly the above suggests two lookups, first, check existence and type. If everything is ok, then do a second lookup to get the actual object.

I have been trying for a solution that only does one MPT access. On asking for a blobber it tries to read the provider into a StorageNode object. 
1. If it actually is a blobber everything ok.
2. If not as many as possable of the fields with matching `json` tags will be copied across. If we make sure all providers have a `json:"provider_type"` field, then this should be filled in. It then makes sense to do the above check:
```go
if blobber.ProviderType != spenum.Blobber {
```

If we are not actually looking at a blobber, the ProviderType field should still be populated as it has the saem `json` tag.

Reply:
Okay, I got what you mean. So your solution is base on the providers objects can be decoded successfully even they are different provider types. For example, if we have a miner of id "abc" and provider type "miner", and saved in MPT, we can load it into a blobber object through `mpt.GetTrieNode()` method, and check its `provider type` field. It may work for us for the providers we now have (not 100 percent sure, need further checking), but that's a little bit hacky as if any of the field of the same name but with different type, the msgp decoding process will fail. The result is not deterministic (at least before we checked all the fields of all the providers).